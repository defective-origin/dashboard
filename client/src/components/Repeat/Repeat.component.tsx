/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react'
import { react } from 'common/tools'

export type ComponentWithItems<OwnProps extends object, T> = OwnProps & {
  items?: T[]
}

export type RepeatVariantProps<
  RC extends Record<string, any>,
> = {
  [key in keyof RC]: { variant?: key } & React.ComponentProps<RC[key]>
}[keyof RC]

export type RepeatComponent<P extends object = any> = React.ElementType<P> | string | Record<string, React.ElementType<P> | string>

export type RepeatItem<
  RC extends RepeatComponent<any>,
> = RC extends React.ElementType<any>
  ? React.ComponentProps<RC>
  : RC extends object ? RepeatVariantProps<RC> : never

export type RepeatItemVariant<
  RC extends RepeatComponent<any>,
> = RC extends object ? RepeatVariantProps<RC>['variant'] : never

export type RepeatProps<
  RC extends RepeatComponent<any> = RepeatComponent<any>,
  RI extends RepeatItem<RC> = RepeatItem<RC>,
  V = RepeatItemVariant<RC>,
> = Partial<RI> & {
  // Component or Component map which should be used for item rendering
  cmp?: RC
  // default variant value
  variant?: V
  // array with item props
  items?: RI[]
  // select key for reconciliation [default: index]
  selectKey?: (item: RI) => React.Key
  // customize props [default: shared props + item props]
  selectProps?: (item: RI) => RI
}

/**
 * Build array with rendered components in it.
 * @example
 * // Render items with one component type
 * function Item({ content }): JSX.Element {
 *   return <p className='item'>{content}</p>
 * }
 *
 * const items: RepeatItem<ItemProps>[] = [
 *   { content: '1' },
 *   { content: '2' },
 *   { content: '3' },
 * ]
 *
 * <Repeat cmp={Item} items={items} shared_propA shared_propB />
 * // <React.Fragment>
 * //   <Item content='1' />
 * //   <Item content='2' />
 * //   <Item content='3' />
 * // </React.Fragment>
 *
 *
 * // Render items with several component types
 * const ItemA = ({ content }) => <p className='itemA'>{content}</p>
 * const ItemB = ({ content }) => <p className='itemB'>{content}</p>
 * const ItemC = ({ content }) => <p className='itemC'>{content}</p>
 *
 * const itemMap = {
 *    a: ItemA,
 *    b: ItemB,
 *    c: ItemC,
 * }
 *
 * const items: RepeatItem<ItemProps>[] = [
 *   { v: 'a', content: '1' },
 *   { v: 'b', content: '2' },
 *   { v: 'c', content: '3' },
 * ]
 *
 * <Repeat cmp={itemMap} items={items} shared_propA shared_propB />
 * // <React.Fragment>
 * //   <ItemA content='1' />
 * //   <ItemB content='2' />
 * //   <ItemC content='3' />
 * // </React.Fragment>
 *
 *
 * // Setup default component type for items
 * <Repeat cmp={itemMap} items={items} v='DEFAULT TYPE' />
 *
 * // Note: You need to cast your items to RepeatItem otherwise typescript can create not appropriate type
 * // if you need to repeat component
 * const items: RepeatItem<typeof Item>[] = []
 *
 * // if you need to repeat components from Map
 * const items: RepeatItem<typeof ITEM_MAP>[] = []
 */
export function Repeat<
  RC extends RepeatComponent<any>,
  RI extends RepeatItem<RC>,
>(props: RepeatProps<RC, RI>): JSX.Element[] | null {
  const { variant: v, cmp, items = [], selectKey, selectProps, ...sharedProps } = props

  if (!cmp || !items.length) {
    return null
  }

  const list = items.map(({ variant = v, ...otherItemProps }, idx) => {
    const Tag = react.isComponent(cmp) ? cmp : cmp[variant]
    const combinedProps = { ...sharedProps, ...otherItemProps } as React.ComponentProps<any>
    const itemProps = selectProps?.(combinedProps) ?? combinedProps
    const itemKey = selectKey?.(itemProps) ?? idx

    return <Tag key={itemKey} {...itemProps} />
  })

  return list
}

Repeat.displayName = 'Repeat'

export default Repeat
