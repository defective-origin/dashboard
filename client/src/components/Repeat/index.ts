export { default } from './Repeat.component'
export * from './Repeat.component'

// export { default } from './RepeatMap.component'
// export * from './RepeatMap.component'
