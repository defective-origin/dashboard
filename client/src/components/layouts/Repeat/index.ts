/* INJECT_IMPORT_PLACE */

/* INJECT_DEFAULT_EXPORT_PLACE */
export { default } from './Repeat.component'

/* INJECT_EXPORT_PLACE */
export * from './Repeat.component'
