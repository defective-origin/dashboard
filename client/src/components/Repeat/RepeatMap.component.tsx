import React from 'react'

// ---| components |---
import Component, { ComponentProps } from 'components/Component'

// ---| self |---
import Repeat, { RepeatItem, RepeatProps } from './Repeat.component'

export type RepeatMapItem<ItemProps extends Record<string, any>> = RepeatItem<ItemProps & ComponentProps>

export type RepeatMapProps<
  ItemMap extends Record<string, any>,
> = RepeatProps<ItemMap[keyof ItemMap] & typeof Component> & {
  // Map of Component which will be used for each array item
  map?: ItemMap
}

/**
 * Build array with rendered components in it.
 * @example
 * type ProxyItemProps = {
 *   text: string
 *   title: string
 *   children?: React.ReactNode
 * }
 *
 * // To do map of components to render by type
 * const componentMap = {
 *  header: () => <div className="header">header</div>
 *  footer: () => <div className="footer">footer</div>
 *  div: () => <div className="div">div</div>
 * }
 *
 * const items: RepeatItem<ProxyItemProps>[] = [
 *   { type: 'header', componentProp: 1 },
 *   { type: 'footer' },
 *   {
 *     type: 'div',
 *     for: [
 *       { type: 'div', },
 *       { type: 'div', },
 *     ],
 *   },
 * ]
 *
 * const renderWithNestedItems = <Repeat map={componentMap} for={items} flatten additional_prop />
 * // // render all nested arrays into first level array (like array.flat(Infinity))
 * // <React.Fragment>
 * //   <div className="header">header</div>
 * //   <div className="footer">footer</div>
 * //   <React.Fragment>
 * //     <div className="div">div</div>
 * //     <React.Fragment>
 * //       <div className="div">div</div>
 * //       <div className="div">div</div>
 * //     </React.Fragment>
 * //   </React.Fragment>
 * // </React.Fragment>
 */
export default function RepeatMap<
  ItemMap extends Record<string, any>,
>(props: RepeatMapProps<ItemMap>): JSX.Element | null {
  return <Repeat item={Component} {...props} />
}

RepeatMap.Items = Repeat
