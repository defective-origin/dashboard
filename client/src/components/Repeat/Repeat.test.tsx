// ---| tests |---
import { render } from 'tests'

// ---| self |---
import Repeat, { RepeatItem } from './Repeat.component'

type ItemProps = { content: React.ReactNode }

describe('[Repeat] component', () => {
  it('should render items with item component', () => {
    const Item = (props: ItemProps) => <p data-testid='item'>{props.content}</p>
    const items: RepeatItem<typeof Item>[] = [
      { content: '1' },
      { content: '2' },
      { content: '3' },
    ]
    const container = render(<Repeat as={Item} items={items} />)

    expect(container.getAll('item')).toHaveLength(3)
    expect(container.snapshot()).toMatchSnapshot()
  })

  it('should render items with item component from map', () => {
    const ItemA = (props: { a: React.ReactNode }) => <p data-testid='itemA'>{props.a}</p>
    const ItemB = (props: { b: React.ReactNode }) => <p data-testid='itemB'>{props.b}</p>
    const ItemC = (props: { c: React.ReactNode }) => <p data-testid='itemC'>{props.c}</p>
    const itemMap = {
      a: ItemA,
      b: ItemB,
      c: ItemC,
    }
    const items: RepeatItem<typeof itemMap>[] = [
      { v: 'a', a: '1' },
      { v: 'b', b: '2' },
      { v: 'c', c: '3' },
    ]
    const container = render(<Repeat as={itemMap} items={items} />)

    expect(container.getAll('itemA')).toHaveLength(1)
    expect(container.getAll('itemB')).toHaveLength(1)
    expect(container.getAll('itemC')).toHaveLength(1)
    expect(container.snapshot()).toMatchSnapshot()
  })

  it('should render items with default component type', () => {
    const ItemA = (props: ItemProps) => <p data-testid='itemA'>{props.content}</p>
    const ItemB = (props: ItemProps) => <p data-testid='itemB'>{props.content}</p>
    const ItemC = (props: ItemProps) => <p data-testid='itemC'>{props.content}</p>
    const itemMap = {
      a: ItemA,
      b: ItemB,
      c: ItemC,
    }
    const items: RepeatItem<typeof itemMap>[] = [
      { content: '1' },
      { content: '2', v: 'b' },
      { content: '3' },
    ]
    const container = render(<Repeat as={itemMap} items={items} v='a' />)

    expect(container.getAll('itemA')).toHaveLength(2)
    expect(container.getAll('itemB')).toHaveLength(1)
    expect(container.snapshot()).toMatchSnapshot()
  })
})
