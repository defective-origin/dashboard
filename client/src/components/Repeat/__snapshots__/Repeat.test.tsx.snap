// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[Repeat] component > should render items with default component type 1`] = `
<DocumentFragment>
  <p
    data-testid="itemA"
  >
    1
  </p>
  <p
    data-testid="itemB"
  >
    2
  </p>
  <p
    data-testid="itemA"
  >
    3
  </p>
</DocumentFragment>
`;

exports[`[Repeat] component > should render items with item component 1`] = `
<DocumentFragment>
  <p
    data-testid="item"
  >
    1
  </p>
  <p
    data-testid="item"
  >
    2
  </p>
  <p
    data-testid="item"
  >
    3
  </p>
</DocumentFragment>
`;

exports[`[Repeat] component > should render items with item component from map 1`] = `
<DocumentFragment>
  <p
    data-testid="itemA"
  >
    1
  </p>
  <p
    data-testid="itemB"
  >
    2
  </p>
  <p
    data-testid="itemC"
  >
    3
  </p>
</DocumentFragment>
`;
